jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java

```diff
@@ -529,6 +529,9 @@ private void registerExternalEventNodeListeners() {
                 List<String> events = ((EventSubProcessNode) node).getEvents();
                 for (String type : events) {
                     addEventListener(type, EMPTY_EVENT_LISTENER, true);
+                    if (isVariableExpression(type)) {
```

Comment:
if isVariableExpression we add two event listeneres - is that right thing to do or should do this check first?

Reply:
yes, that's on purpose to allow variable expression to be resolved on demand - for instance when process instance is going to be persisted, duplicates will be removed as all events are kept in set.

Reply:
Ok. thanks for explaining

Reply:
Hmm, so we have one listener with the current value of variable and the second one with unresolved expression? When we update the variable, the old one get's removed or not?

Reply:
these two are only registered for the time process instance is being executed so once it's done (tx is being committed) it will the store only one.